### PR TITLE
Add a notebook to create entry point for training models

### PR DESCRIPTION
**Pull request by abcdefgs0324**
- Add a notebook to create entry point for training models with fire
- Filter labels by frequency
- Load embeddings from GCS and push trained models to GCS
- Write MLP classifier as a module
- Fix Label_k8s notebook errors

/assign @jlewi 
/assign @zhenghuiwang 
/assign @hamelsmu

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/code-intelligence/20)

### CODE REVIEW

File: Issue-Label-Bot-v2/notebooks/config.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+import os
+import yaml
+
+class Config(object):
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/notebooks/config.py]:
```suggestion
class Config:
```

**Comment by hamelsmu** [comment]
@jlewi can you please install [this GitHub app](https://github.com/marketplace/review-notebook-app) on this repo to facilitate review of notebooks?

**Comment by hamelsmu** [comment]
It is odd that Recall is not 100% when the threshold is zero.  This might be a bug

image

File: Issue-Label-Bot-v2/notebooks/mlp.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+        self.precision = pred_correct / total_pred_true
+        self.recall = pred_correct / total_true
+
+        df = pd.DataFrame({'label': self.label_columns,
+                           'precision': self.precision,
+                           'recall': self.recall})
+        return df, self.precision, self.recall
+
+    def find_best_prob_thre(self, y_true, y_pred):
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/notebooks/mlp.py]: I'm really confused about this function.  The precision and recall threshold should be supplied by the user, as there isn't an objective way to automatically choose the best precision and recall thresholds automatically

**Comment by abcdefgs0324** [comment]
@hamelsmu 
In the function `calculate_argmax_result`, the parameter `threshold` means the threshold that the **highest** probability of labels should be larger than, and then **the highest label** will be predicted. So, setting `threshold=0` does not means I return all labels to achieve 100% recall because currently I only return the label with the highest probability for each prediction.

File: Issue-Label-Bot-v2/notebooks/mlp.py (hunk 2 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: In this function, I would like to find a threshold which achieves more number of labels with precision larger than the predefined threshold, precision_thre.
If two thresholds have the same number of labels with precision larger than the precision threshold, I would choose the one with more labels having recall larger than the recall threshold.
That is because users may be more sensitive to false positive error and the recall is quite low when we set a higher threshold as I only predict one label at this moment.
This is not a good way to find the best threshold and we may need to figure out how to set a threshold to predict multiple labels instead of only one label.
If you are still confused, please let me know and we can have a phone call. Thanks.

File: Issue-Label-Bot-v2/notebooks/config.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+        all_config['model_local_path'] = f'{repo_name}.dpkl'
+        all_config['labels_local_path'] = f'{repo_name}.labels.dpkl'
+        all_config['emb_local_path'] = f'{repo_name}.emb.dpkl'
+
+        return all_config
+
+
+if __name__ == '__main__':
+    config = Config(yaml_path='issue_label_bot.yaml').load()
```
Comment (zhenghuiwang) [inline Issue-Label-Bot-v2/notebooks/config.py]: This API is unusual. Normally you will have `config = Config(yaml_path=...)` or `config = load_config(...)`.

If you define `Config` as a class, all the properties of `all_config` should be defined as the properties of `Config` and set in the `__init__` method.

You could load the dictionary from YAML and use `self.__dict__.update()` to update corresponding properties.

File: Issue-Label-Bot-v2/notebooks/mlp.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+from sklearn.neural_network import MLPClassifier
+from sklearn.model_selection import GridSearchCV
+from sklearn.metrics import roc_auc_score
+import dill as dpickle
+import numpy as np
+import pandas as pd
+
+
+class MLP:
```
Comment (zhenghuiwang) [inline Issue-Label-Bot-v2/notebooks/mlp.py]: Can you add some test cases to test this class and also demonstrate how to use it?

File: Issue-Label-Bot-v2/notebooks/config.py (hunk 3 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/config.py]: I will write all as the properties of `Config`.

File: Issue-Label-Bot-v2/notebooks/mlp.py (hunk 4 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: Okay!

File: Issue-Label-Bot-v2/notebooks/mlp.py (hunk 2 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: I will update to predict multiple labels with the predefined threshold, which much makes sense. So, the recall when `threshold=0.0` should be 100%.

Comment (hamelsmu) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: @abcdefgs0324 just to recap, this is the way I was thinking to manage the thresholds:

1.  Have the user supply their desired precision.  Default: 70%
2.  Have the user supply their minimum recall at this precision. Default 50%
3.  Find the threshold that meets this minimum standard.  There is a precision-recall tradeoff so the higher the precision the lower the recall these two metrics are strictly inversely related.   **This is why I'm confused why there is a search for the best threshold, as it is deterministic based on the precision / recall that the user chooses** . Keep in mind that the user will rarely choose this and the defaults will probably control 99% of the time.  

Let me know your thoughts or where I might be missing something.  Thanks!

Comment (hamelsmu) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: Also can you clarify what you mean when you say you are predicting only one label?  Can you shed some light on why?  I have a hunch that some of the confusion has to do with that

Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: I have updated the code to predict multiple labels. The procedure is that we set a threshold `prob_thre`, and all labels with probabilities larger than this threshold will be predicted `True`. You can see [the function `calculate_result`](https://github.com/abcdefgs0324/Issue-Label-Bot/blob/22dcf6ee2acf5810262f5745cee1e24a20c39199/Issue-Label-Bot-v2/notebooks/mlp.py#L91) in `mlp.py`.

Yes, the user will supply two values for precision and recall. And this function `find_best_prob_thre` is to find the threshold that meets precision and recall standards. However, there is a problem that we can not have a threshold to let all labels meet both of the precision and recall minimum.
You can see `Label_k8s_issues_with_MLP.ipynb` (cell 45), when `prob_thre=0.0`, some labels such as `sig/vmware` and `sig/service-catalog` have quite low precision. And it is impossible to let these labels to meet the precision minimum. So in my current implementation, I find a threshold to let more number of labels to meet both of precision and recall minimum and this is what this function does.

For example, assume when `prob_thre=0.5`, we have 10 labels that meet both of minimum. And if increasing the `prob_thre` only decreases the number of labels meeting two minimum, this function will choose `prob_thre=0.5` to be its best threshold.

**Comment by jlewi** [comment]
Link to review with reviewnb
https://app.reviewnb.com/kubeflow/code-intelligence/pull/20/files/

File: Issue-Label-Bot-v2/notebooks/mlp.py (hunk 2 above)
Comment (hamelsmu) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: > So in my current implementation, I find a threshold to let more number of labels to meet both of precision and recall minimum and this is what this function does.

What do you mean by more number of labels?  more than what?  I was expecting that if labels do not meet the precision / threshold criteria you exclude them all together and do not offer any predictions on those labels, since the model would create more noise than value.  

> For example, assume when prob_thre=0.5, we have 10 labels that meet both of minimum.

I find it confusing that `prob_thre` is a parameter.  prob_thre should be something you calculate internally from the precision and recall thresholds, but **not** something that is specified by the user.  Perhaps this is an additional source of confusion?  

The api that expect is
```py
# set desired precision and recall
estimator.set_precision_recall(precision=<float>, recall=<float>)

# your estimator will return a list of labels.  If a label cannot be predicted with 
# precision/recall that means that your model is not good enough to opine on those 
# labels and should simply ignore those particular labels.
estimator.predict(body:str, title:str) -> List
```

Also, please consider using [sklearn's built in utility](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.precision_recall_curve.html) for finding the proper threshold if possible.  That might simplify the code and make your life easier!

**Comment by hamelsmu** [comment]
@abcdefgs0324 Consider dropping the part of the notebook that caluclates AUC, as AUC was useful for comparing different models to eachother.  In this case, you are only running one model so I think this part of the notebbook is extreanous

**Comment by hamelsmu** [comment]
@abcdefgs0324 just want to say I'm just writing freely and don't mean anything in a harsh way.  We are throwing lots of suggestions at you and it is definitely challenging area covering many disciplines (ML, devops, etc).  Thanks for taking our feedback.  I know that getting lots of feedback on a PR can be intimidating, but rest assured only giving this feedback with ❤️  ... Thanks for working on this project!

**Comment by hamelsmu** [comment]
@jlewi is there a CI system that Kubeflow likes to use?  Perhaps @abcdefgs0324 might like to add that (or I can help) as he has written tests for this code and would be nice to turn on automated testing

File: Issue-Label-Bot-v2/tests/test_mlp.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
+import pytest
+import sys
+sys.path.insert(1, '../notebooks')
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/tests/test_mlp.py]: If you want you can add a symlink and it will work on GitHub as well that way you don't have to worry about changing your python path.  Not necessary just a suggestion to make your life easier

File: Issue-Label-Bot-v2/tests/test_mlp.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    mlp_clf.fit(X_train, y_train)
+    mlp_clf_pred = mlp_clf.predict_proba(X_test)
+
+    mlp_wrap = MLPWrapper(clf=mlp_clf)
+    mlp_wrap.fit(X_train, y_train)
+    mlp_wrap_pred = mlp_wrap.predict_proba(X_test)
+    assert mlp_clf_pred.all() == mlp_wrap_pred.all()
+
+
+def test_find_probability_thresholds():
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/tests/test_mlp.py]: thanks for adding tests

File: Issue-Label-Bot-v2/tests/test_mlp.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+    random_state = 1234
+    precision_threshold = 0.7
+    recall_threshold = 0.5
+
+    mlp_clf = MLPClassifier(random_state=random_state)
+
+    mlp_wrap = MLPWrapper(clf=mlp_clf, precision_threshold=precision_threshold,
+                          recall_threshold=recall_threshold)
+    mlp_wrap.find_probability_thresholds(X, y)
+    thresholds = mlp_wrap.probability_thresholds
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/tests/test_mlp.py]: Please add a comment on what `thresholds` is?  Is it a Tuple(Float, Float, Float) ?

File: Issue-Label-Bot-v2/tests/test_mlp.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    precision_0, recall_0 = mlp_wrap.precisions[0], mlp_wrap.recalls[0]
+    assert not thresholds[1] and not thresholds[2] and\
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/tests/test_mlp.py]: Why are thresholds for the last two labels supposed to be None?  This seems counter-intuitive to me ....  

Also why are we only testing the thresholds from the first label indexed at [0] ?

File: Issue-Label-Bot-v2/notebooks/mlp.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
+from sklearn.neural_network import MLPClassifier
+from sklearn.model_selection import GridSearchCV
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import precision_recall_curve
+import dill as dpickle
+import numpy as np
+import pandas as pd
+import logging
+
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/notebooks/mlp.py]: please add requirements.txt file with strictly pinned versions as will make it easier to reproduce your environment.

File: Issue-Label-Bot-v2/notebooks/Repo_MLP.ipynb [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+   "source": [
+    "r.train()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Create entry point using fairing"
+   ]
```
Comment (hamelsmu) [inline Issue-Label-Bot-v2/notebooks/Repo_MLP.ipynb]: Do you mind explaining what the below cell is doing as an explanation in the notebook?  This would be great for anyone who is browsing Kubeflow and wants to understand what is happening looking at this notebook.   Also would be good to explain what the fairing markers are doing in earlier cells and what functionality they provide

Comment (hamelsmu) [reply Issue-Label-Bot-v2/notebooks/Repo_MLP.ipynb]: I dug around for documentation on Fairing and it does not exist or is extremely sparse, so anything you can add here in terms of explanation and how it is working will be extremely helpful!

Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/Repo_MLP.ipynb]: Sure, I will do that

File: Issue-Label-Bot-v2/notebooks/mlp.py (hunk 9 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/notebooks/mlp.py]: Okay

File: Issue-Label-Bot-v2/tests/test_mlp.py (hunk 7 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/tests/test_mlp.py]: Ok

File: Issue-Label-Bot-v2/tests/test_mlp.py (hunk 8 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/tests/test_mlp.py]: Well, for this test case, I want to make sure the probability thresholds are correct. Therefore, I check the output of the MLP classifier from sklearn (not my wrapper) and know what the correct thresholds for this case are.
Any suggestion to do something like this? I seldom write unit tests...

The probability thresholds of all the labels which do not meet precision and recall thresholds (those excluded labels) will be `None`. I will add comments on this.

File: Issue-Label-Bot-v2/tests/test_mlp.py (hunk 5 above)
Comment (abcdefgs0324) [reply Issue-Label-Bot-v2/tests/test_mlp.py]: That sounds good. I will have a look.

File: Issue-Label-Bot-v2/tests/test_mlp.py (hunk 8 above)
Comment (hamelsmu) [reply Issue-Label-Bot-v2/tests/test_mlp.py]: Oh I understand you are checking your code against sklearn as the reference?  I didn't realize that is what you were intending to do, that makes sense.

**Comment by jlewi** [comment]
We use prow to run our tests (unit & E2E). There's some documentation [here](https://github.com/kubeflow/testing) about setting up prow.

**Comment by jlewi** [comment]
Thanks @abcdefgs0324 I left a couple minor comments Thanks for persevering through all our comments. 

On my end this PR is LGTM; so I think as soon as @hamelsmu comments are addressed we can merge it.

**Comment by jlewi** [comment]
/lgtm
/approve

But it looks like you need to sync and resolve the conflicts.

**Comment by abcdefgs0324** [comment]
Resolved the conflicts.
